### PR TITLE
Simplify the associated types for Computed State & Sub States

### PR DESCRIPTION
**Pull request by lee-orr**
# Objective
The goal is to simplify the associated types of the ComputedState and SubState traits, since otherwise they end up producing hard to parse arguments for the `compute` or `exists` functions: such as `<<GameState as ComputedState>::SourceStates as StateSet>::OptionalStateSet`

## Solution

Add a private `InnerStateSet` type that is implemented for both `S: State` and `Option<S: State>`. Implement `StateSet` for `S: InnerStateSet` (and tuples of it) - if `S` isn't optional and doesn't exist in the world when `compute` or `exists` would be called, it is treated as if it returned None.

### CODE REVIEW

File: crates/bevy_ecs/src/schedule/state.rs [hunk 1]
```diff
@@ -579,19 +573,50 @@ pub trait StateSet: sealed::StateSetSealed {
     );
 }
 
-impl<S: States> StateSetSealed for S {}
+trait InnerStateSet: Sized {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/state.rs:589]: Doc comments on what this is and why it's useful please :) Very easy to get confused about this design in the future.

**Comment by alice-i-cecile** [comment]
Good, I like this!

File: crates/bevy_ecs/src/schedule/state.rs (hunk 1 above)
Comment (lee-orr) [reply crates/bevy_ecs/src/schedule/state.rs:589]: Done

**Comment by lee-orr** [comment]
Seeing both @alice-i-cecile's comment here and @james7132's comment on Discord - merging this in to the actual PR
